app/controllers/api/v1/questionnaires_controller.rb

```diff
@@ -0,0 +1,169 @@
+class QuestionnairesController < ApplicationController
+  include AuthorizationHelper
+
+  # Controller for Questionnaire objects
+  # A Questionnaire can be of several types (QuestionnaireType)
+  # Each Questionnaire contains zero or more questions (Question)
+  # Generally a questionnaire is associated with an assignment (Assignment)
+
+  
+  skip_before_action :verify_authenticity_token
+
+  # Check role access for edit questionnaire
+  def action_allowed?
+    case params[:action]
+    when 'edit'
+      @questionnaire = Questionnaire.find(params[:id])
+      current_user_has_admin_privileges? ||
+        (current_user_is_a?('Instructor') && current_user_id?(@questionnaire.try(:instructor_id))) ||
+        (current_user_is_a?('Teaching Assistant') && session[:user].instructor_id == @questionnaire.try(:instructor_id))
+    else
+      current_user_has_student_privileges?
+    end
+  end
+
+  #List all the questionnaires
+  # GET on /questionnaires
+  def index
+    @questionnaires = Questionnaire.order(:id)
+    render json: @questionnaires
+  end
+  
+  #Show a given questionnaire
+  # GET on /questionnaires/:id
+  def show
+    begin
+      @questionnaire = Questionnaire.find(params[:id])
+      render json: @questionnaire
+    rescue
+      msg = "No such Questionnaire exists."
+      render json: msg, status: :not_found
+    end
+  end
+  
+  #Create a given questionnaire
+  # POST on /questionnaires
+  def create
+    if params[:questionnaire][:name].blank?
+      redirect_to controller: 'questionnaires', action: 'new', model: params[:questionnaire][:type], private: params[:questionnaire][:private]
+    else
+      questionnaire_private = params[:questionnaire][:private] == 'true'
+      display_type = params[:questionnaire][:type].split('Questionnaire')[0]
+      begin
+        @questionnaire = Object.const_get(params[:questionnaire][:type]).new if Questionnaire::QUESTIONNAIRE_TYPES.include? params[:questionnaire][:type]
+      rescue StandardError
+        msg = $ERROR_INFO
+        render json: msg
+      end
+      begin
+        @questionnaire.private = questionnaire_private
+        @questionnaire.name = params[:questionnaire][:name]
+        @questionnaire.instructor_id = 6 # session[:user].id
+        @questionnaire.min_question_score = params[:questionnaire][:min_question_score]
+        @questionnaire.max_question_score = params[:questionnaire][:max_question_score]
+        @questionnaire.type = params[:questionnaire][:type]
+        if %w[AuthorFeedback CourseSurvey TeammateReview GlobalSurvey AssignmentSurvey BookmarkRating].include?(display_type)
+          display_type = (display_type.split(/(?=[A-Z])/)).join('%')
+        end
+        @questionnaire.display_type = display_type
+        @questionnaire.instruction_loc = Questionnaire::DEFAULT_QUESTIONNAIRE_URL
+        @questionnaire.save
+        # Create node
+        tree_folder = TreeFolder.where(['name like ?', @questionnaire.display_type]).first
```

Comment:
Now that your working on the front end, analyze that do we really need tree structure in rails or we can handle that in front-end

Reply:
Commit [f039d84](https://github.com/expertiza/reimplementation-back-end/pull/20/commits/f039d843a0176aa490bc542fe0cf6d42cca62d49) addresses this review